Tools/cases_generator/generate_cases.py

```diff
@@ -107,46 +133,56 @@ def write_instr(instr: InstDef, predictions: set[str], indent: str, f: TextIO, d
                 f.write(f"{space}if ({cond}) goto {label};\n")
         else:
             f.write(line)
-    noutputs = len(instr.outputs or ())
+    if always_exits(instr.block):
+        # None of the rest matters
+        return cache_offset
+    # Stack effect
+    noutputs = len(outputs)
     diff = noutputs - ninputs
     if diff > 0:
         f.write(f"{indent}    STACK_GROW({diff});\n")
     elif diff < 0:
         f.write(f"{indent}    STACK_SHRINK({-diff});\n")
-    for i, output in enumerate(reversed(instr.outputs or ()), 1):
-        if output not in (instr.inputs or ()):
-            f.write(f"{indent}    POKE({i}, {output});\n")
-    assert instr.block
-
-def write_cases(f: TextIO, instrs: list[InstDef], supers: list[parser.Super]):
+    for i, output in enumerate(reversed(outputs), 1):
+        if output.name not in input_names and output.name != "unused":
+            f.write(f"{indent}    POKE({i}, {output.name});\n")
+    # Cache effect
+    if cache_offset:
+        f.write(f"{indent}    next_instr += {cache_offset};\n")
+    return cache_offset
+
+
+def write_cases(
+    f: TextIO, instrs: list[InstDef], supers: list[parser.Super], families: list[parser.Family]
+) -> dict[str, tuple[int, int, int]]:
     predictions: set[str] = set()
     for instr in instrs:
-        assert isinstance(instr, InstDef)
-        assert instr.block is not None
         for target in re.findall(RE_PREDICTED, instr.block.text):
             predictions.add(target)
     indent = "        "
     f.write(f"// This file is generated by {os.path.relpath(__file__)}\n")
     f.write(f"// Do not edit!\n")
     instr_index: dict[str, InstDef] = {}
+    effects_table: dict[str, tuple[int, int, int]] = {}  # name -> (ninputs, noutputs, cache_offset)
     for instr in instrs:
         instr_index[instr.name] = instr
         f.write(f"\n{indent}TARGET({instr.name}) {{\n")
         if instr.name in predictions:
             f.write(f"{indent}    PREDICTED({instr.name});\n")
-        write_instr(instr, predictions, indent, f)
-        assert instr.block
+        cache_offset = write_instr(
+            instr, predictions, indent, f,
+            cache_size=find_cache_size(instr, families)
+        )
+        effects_table[instr.name] = len(instr.inputs), len(instr.outputs), cache_offset
```

Comment:
Hm. This is a bit weird because it treats caches and stack items the same, which doesn't make much sense. It seems to me we'd want something that captures just stack effect and cache size:

```suggestion
        stack_pre = sum(isinstance(item, StackEffect) for item in instr.inputs)
        stack_post = sum(isinstance(item, StackEffect) for item in instr.inputs)
        effects_table[instr.name] = stack_post - stack_pre, cache_offset
```

(This will probably get more complicated as stack effects start getting more complicated...)